File: internal/auction/engine.go
package auction

import (
	"context"
	"database/sql"
	"errors"
	"fmt"
	"log/slog"
	"sync"
	"time"

	"github.com/vgartg/goauction/internal/metrics"
	"github.com/vgartg/goauction/internal/models"
	"github.com/vgartg/goauction/internal/repository"
)

type WSBroadcaster interface {
	BroadcastToLot(lotID string, message interface{})
}

type Config struct {
	// Anti-sniping: if a bid lands within SnipingWindow before close,
	// extend ClosingAt by SnipingExtension. Zero window disables it.
	SnipingWindow    time.Duration
	SnipingExtension time.Duration
}

type Engine struct {
	repo repository.LotRepository
	ws   WSBroadcaster
	cfg  Config

	timersMu sync.Mutex
	timers   map[string]*time.Timer
}

func NewEngine(repo repository.LotRepository, ws WSBroadcaster, cfg Config) *Engine {
	return &Engine{
		repo:   repo,
		ws:     ws,
		cfg:    cfg,
		timers: make(map[string]*time.Timer),
	}
}

func (e *Engine) CreateLot(ctx context.Context, title string, startPrice, minStep float64, closingAt time.Time) (*models.Lot, error) {
	lot := &models.Lot{
		Title:        title,
		StartPrice:   startPrice,
		MinStep:      minStep,
		CurrentPrice: startPrice,
		Status:       models.LotStatusActive,
		ClosingAt:    closingAt,
		Version:      1,
	}
	if err := e.repo.CreateLot(ctx, lot); err != nil {
		return nil, err
	}
	metrics.ActiveLots.Inc()
	e.scheduleClose(lot)
	return lot, nil
}

func (e *Engine) GetLot(ctx context.Context, id string) (*models.Lot, error) {
	return e.repo.GetLotByID(ctx, id, false)
}

func (e *Engine) ListLots(ctx context.Context) ([]*models.Lot, error) {
	return e.repo.GetAllLots(ctx)
}

func (e *Engine) RecentBids(ctx context.Context, lotID string, limit int) ([]*models.Bid, error) {
	return e.repo.GetRecentBids(ctx, lotID, limit)
}

func (e *Engine) PlaceBid(ctx context.Context, lotID, userID string, amount float64) (*models.Lot, error) {
	for attempts := 0; attempts < 3; attempts++ {
		lot, err := e.repo.GetLotByID(ctx, lotID, true)
		if err != nil {
			return nil, err
		}
		if lot.Status != models.LotStatusActive {
			return nil, errors.New("lot is not active")
		}
		if time.Now().After(lot.ClosingAt) {
			return nil, errors.New("lot already closed")
		}
		if amount <= lot.CurrentPrice {
			return nil, fmt.Errorf("bid must be higher than current price %.2f", lot.CurrentPrice)
		}
		if amount < lot.CurrentPrice+lot.MinStep {
			return nil, fmt.Errorf("bid must be at least %.2f more than current price", lot.MinStep)
		}

		bid := &models.Bid{
			LotID:  lotID,
			UserID: userID,
			Amount: amount,
		}
		if err := e.repo.CreateBid(ctx, bid); err != nil {
			return nil, err
		}

		oldVersion := lot.Version
		lot.CurrentPrice = amount
		lot.Version++

		// Anti-sniping: extend close window if bid lands near the end.
		extended := false
		if e.cfg.SnipingWindow > 0 && time.Until(lot.ClosingAt) <= e.cfg.SnipingWindow {
			lot.ClosingAt = time.Now().Add(e.cfg.SnipingExtension)
			lot.ExtendedCount++
			extended = true
		}

		if err := e.repo.UpdateLot(ctx, lot, oldVersion); err != nil {
			if errors.Is(err, repository.ErrOptimisticLock) {
				continue
			}
			return nil, err
		}

		metrics.BidsTotal.WithLabelValues(lotID).Inc()

		e.ws.BroadcastToLot(lotID, map[string]interface{}{
			"type":      "new_bid",
			"lot_id":    lotID,
			"user_id":   userID,
			"amount":    amount,
			"new_price": amount,
			"timestamp": time.Now().UTC(),
		})

		if extended {
			e.scheduleClose(lot)
			e.ws.BroadcastToLot(lotID, map[string]interface{}{
				"type":           "lot_extended",
				"lot_id":         lotID,
				"closing_at":     lot.ClosingAt.UTC(),
				"extended_count": lot.ExtendedCount,
			})
		}
		return lot, nil
	}
	return nil, errors.New("failed to place bid after retries")
}

func (e *Engine) CloseLot(lot *models.Lot) error {
	ctx := context.Background()
	if lot.Status != models.LotStatusActive {
		return nil
	}
	if time.Now().Before(lot.ClosingAt) {
		// Closing was extended after this timer was scheduled. Reschedule and bail.
		e.scheduleClose(lot)
		return nil
	}
	highestBid, err := e.repo.GetHighestBid(ctx, lot.ID)
	if err != nil && !errors.Is(err, sql.ErrNoRows) {
		return err
	}
	if highestBid != nil {
		lot.WinnerID = &highestBid.UserID
	}
	oldVersion := lot.Version
	lot.Status = models.LotStatusClosed
	lot.Version++
	if err := e.repo.UpdateLot(ctx, lot, oldVersion); err != nil {
		return err
	}
	metrics.LotClosuresTotal.Inc()
	metrics.ActiveLots.Dec()
	e.clearTimer(lot.ID)
	e.ws.BroadcastToLot(lot.ID, map[string]interface{}{
		"type":        "lot_closed",
		"lot_id":      lot.ID,
		"winner_id":   lot.WinnerID,
		"final_price": lot.CurrentPrice,
	})
	return nil
}

// StartTimerForLot is used at boot to restore timers for active lots.
func (e *Engine) StartTimerForLot(lot *models.Lot) {
	e.scheduleClose(lot)
}

// scheduleClose installs (or replaces) the auto-close timer for a lot.
func (e *Engine) scheduleClose(lot *models.Lot) {
	duration := time.Until(lot.ClosingAt)
	if duration <= 0 {
		go func() {
			if err := e.closeLotByID(lot.ID); err != nil {
				slog.Error("failed to close lot", "lot_id", lot.ID, "error", err)
			}
		}()
		return
	}
	timer := time.AfterFunc(duration, func() {
		if err := e.closeLotByID(lot.ID); err != nil {
			slog.Error("failed to close lot", "lot_id", lot.ID, "error", err)
		}
	})
	e.timersMu.Lock()
	if old, ok := e.timers[lot.ID]; ok {
		old.Stop()
	}
	e.timers[lot.ID] = timer
	e.timersMu.Unlock()
}

func (e *Engine) clearTimer(lotID string) {
	e.timersMu.Lock()
	defer e.timersMu.Unlock()
	if t, ok := e.timers[lotID]; ok {
		t.Stop()
		delete(e.timers, lotID)
	}
}

func (e *Engine) closeLotByID(lotID string) error {
	ctx := context.Background()
	fresh, err := e.repo.GetLotByID(ctx, lotID, false)
	if err != nil {
		return err
	}
	if fresh.Status != models.LotStatusActive {
		return nil
	}
	return e.CloseLot(fresh)
}

File: web/src/engine.ts
import { emit } from './events';
import { store } from './store';
import type { Bid, Lot } from './types';
import { uuid } from './uuid';

export const SNIPING_WINDOW_MS = 30_000;
export const SNIPING_EXTENSION_MS = 30_000;

const timers = new Map<string, ReturnType<typeof setTimeout>>();

export function createLot(input: {
  title: string;
  start_price: number;
  min_step: number;
  closing_at: Date;
}): Lot {
  if (!input.title.trim()) throw new Error('title is required');
  if (input.start_price <= 0) throw new Error('start price must be positive');
  if (input.min_step <= 0) throw new Error('min step must be positive');
  if (input.closing_at.getTime() <= Date.now()) throw new Error('closing time must be in the future');

  const lot: Lot = {
    id: uuid(),
    title: input.title.trim(),
    start_price: input.start_price,
    min_step: input.min_step,
    current_price: input.start_price,
    status: 'active',
    created_at: new Date().toISOString(),
    closing_at: input.closing_at.toISOString(),
    version: 1,
    winner_id: null,
    extended_count: 0,
  };
  store.upsertLot(lot);
  scheduleClose(lot);
  emit({ type: 'lot_changed', lot });
  return lot;
}

export function placeBid(lotID: string, userID: string, amount: number): Lot {
  const lot = store.getLot(lotID);
  if (!lot) throw new Error('lot not found');
  if (lot.status !== 'active') throw new Error('lot is not active');
  if (Date.now() > new Date(lot.closing_at).getTime()) throw new Error('lot already closed');
  if (amount <= lot.current_price) {
    throw new Error(`bid must be higher than current price ${lot.current_price.toFixed(2)}`);
  }
  if (amount < lot.current_price + lot.min_step) {
    throw new Error(`bid must be at least ${lot.min_step.toFixed(2)} more than current price`);
  }

  const bid: Bid = {
    id: uuid(),
    lot_id: lotID,
    user_id: userID,
    amount,
    created_at: new Date().toISOString(),
  };
  store.addBid(bid);

  lot.current_price = amount;
  lot.version += 1;

  let extended = false;
  const msToClose = new Date(lot.closing_at).getTime() - Date.now();
  if (msToClose <= SNIPING_WINDOW_MS) {
    lot.closing_at = new Date(Date.now() + SNIPING_EXTENSION_MS).toISOString();
    lot.extended_count += 1;
    extended = true;
  }
  store.upsertLot(lot);

  emit({ type: 'new_bid', lot_id: lotID, bid, new_price: amount });
  emit({ type: 'lot_changed', lot });
  if (extended) {
    scheduleClose(lot);
    emit({
      type: 'lot_extended',
      lot_id: lotID,
      closing_at: lot.closing_at,
      extended_count: lot.extended_count,
    });
  }
  return lot;
}

export function closeLot(lotID: string): void {
  const lot = store.getLot(lotID);
  if (!lot || lot.status !== 'active') return;
  if (Date.now() < new Date(lot.closing_at).getTime()) {
    scheduleClose(lot);
    return;
  }
  const top = store.highestBid(lotID);
  if (top) lot.winner_id = top.user_id;
  lot.status = 'closed';
  lot.version += 1;
  store.upsertLot(lot);
  clearTimer(lotID);
  emit({
    type: 'lot_closed',
    lot_id: lotID,
    winner_id: lot.winner_id,
    final_price: lot.current_price,
  });
  emit({ type: 'lot_changed', lot });
}

export function scheduleClose(lot: Lot): void {
  clearTimer(lot.id);
  const ms = new Date(lot.closing_at).getTime() - Date.now();
  if (ms <= 0) {
    queueMicrotask(() => closeLot(lot.id));
    return;
  }
  const t = setTimeout(() => closeLot(lot.id), Math.min(ms, 2_147_483_000));
  timers.set(lot.id, t);
}

function clearTimer(lotID: string): void {
  const t = timers.get(lotID);
  if (t) {
    clearTimeout(t);
    timers.delete(lotID);
  }
}

export function restoreTimers(): void {
  for (const lot of store.listLots('active')) scheduleClose(lot);
}
